### ISSUE TITLE
Unable to run large jar generated by lein-binplus

### ISSUE THREAD

**Issue by cross** [BrunoBonacci/lein-binplus#5]
We were successfully generating uberjars that were wrapped by lein-binplus successfully.  But, recent work done by a member of my team increased the number of classes in this jar above 65536, and reportedly Java 7 cannot run a jar by manifest if the jar exceeds that limit.

The uberjar now being generated _can_ be run by "java ${UBERJAR} ${main}", but putting the same preamble into a generate bin with lein-binplus does not work.  And, running "java ${binplus-jar} ${main}" fails in the same way.  The error reported is:

```
Failure msg: Error occurred during initialization of VM
java/lang/ClassNotFoundException: error in opening JAR file <Zip file open error> _binplus-jar_
```

Is this a problem that _can_ be solved?  At the moment, my developer has failed to find an answer, and we're switching to an external shell script to invoke the uber-jar directly.  But, I'd rather we can reduce it back down to one executable.

**Comment by cross** [comment]
Update.  So, the old binplus binary jar shows the following when viewed with the "jar" command:
```
% jar tvf /bin/ourprogram  | wc
  59234  473873 5570006
%
```
But, the updated binplus-generated binary (which unzip -l shows me contains ~68000 lines of output) shows:
```
% jar tvf ourprogram/target/ourprogram  | wc
java.util.zip.ZipException: invalid END header (bad central directory size)
	at java.util.zip.ZipFile.open(Native Method)
	at java.util.zip.ZipFile.<init>(ZipFile.java:215)
	at java.util.zip.ZipFile.<init>(ZipFile.java:145)
	at java.util.zip.ZipFile.<init>(ZipFile.java:116)
	at sun.tools.jar.Main.list(Main.java:1058)
	at sun.tools.jar.Main.run(Main.java:246)
	at sun.tools.jar.Main.main(Main.java:1231)
      0       0       0
%
``` 

I am not a java expert by any stretch.  But, this looks like the issue.  It's that java isn't able to cope with there _being_ a preamble for this larger uberjar.  I don't know if this is surmountable, and look forward to hearing your thoughts.  Thanks.

**Comment by BrunoBonacci** [comment]
Hi,
Could you please send me the exact configuration you have in project.clj?
Could you please also send me the shell script used to execute the program externally?

Bye
Bruno

**Comment by BrunoBonacci** [comment]
From what i can read the only issue is that your custom preamble is incorrect.
Try with this:

```
    :bin {:name "runme"
          :custom-preamble "#!/bin/sh\nexec java {{{jvm-opts}}} -cp $0 {{main}} \"$@\"\n"}
```

**Comment by cross** [comment]
I had a set of :jvm-opts defined within the :bin section of project.clj, then had the following custom-preamble:
`     :custom-preamble "#!/bin/sh\nexec java {{&jvm-opts}} -D{{&project-name}}.version={{&version}} -Xbootclasspath/a:$0 {{&main}} \"$@\"\n"`

As you suggested, I removed the specific `:jvm-opts` I'd set, and have only the two lines you have above.  That has a slightly different failure:

```
% ./runme
Error: Could not find or load main class foo.core
% head -n 2 runme
#!/bin/sh
exec java -server -Dfile.encoding=UTF-8 -cp $0 foo.core "$@"
% 
```

**Comment by cross** [comment]
The shell script, with only altered jar path, is:
```
#!/usr/bin/env bash

FOOJAR=/path_to_our_java_libs/foo-0.3.1-SNAPSHOT-standalone.jar

java -client -XX:+TieredCompilation -XX:TieredStopAtLevel=1 -XX:-OmitStackTraceInFastThrow -cp ${FOOJAR} foo.core "$@"
```

Those options are from the lein-binplus options I had been using before, thus were duplicated in the script.  But merely running "java -cp $FOOJAR" works, but "java -cp $FOOBIN" does not.  Where FOOBIN is the result of lein-binplus on the same uberjar.

**Comment by BrunoBonacci** [comment]
here we might be hitting some inherent limit. I was reading a bit around like:

http://docs.oracle.com/javase/7/docs/technotes/guides/jar/jar.html
https://en.wikipedia.org/wiki/Zip_(file_format)#Design

and they all mention limits on 65535 entries per file.

It is hard for me to test this as I do not have such large project to try a few things.
one last attempt to do might be the following:

  * create uberjar
  * ensure you can start with `java -cp jarname your.main.class`
  * create a shell script called `preamble.sh` with the preamble you'd like 
    * remember to put `$0` instead of the jar name
    * remember to add an additional new line at the end of the file
  * finally run `cat preamble.sh target/uberjar.jar > myexec` and try to run the file

if this doesn't run I'm afraid there is little we can do.
One thing to note is that the large number of classes might cause other secondary issues
like the excessive loading and unloading of classes by the ClassLoader and the need of increasing the PermGen space (if you are running on Java7).

**Comment by cross** [comment]
Confirmed.  Thanks.  Yes, as expected it was that we'd passed the 65k elements.  (We have 68000 ish now).  We had 55k or so before when we were testing this, and I did the work required to eliminate the need for the external shell wrapper (due to opensshd issues).

Oh well.  Thank you for your help.  The jar runs fine if not prefixed with shell, but fails to run itself or be used as an argument to "java -cp" with a prefix of any sort.  Thank you for taking the time to help me look at options.
